http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/MultiplexedChannelRecord.java

```diff
@@ -31,165 +27,194 @@
 
-    /**
-     * @param connectionFuture Future for parent socket channel.
-     * @param maxConcurrencyPerConnection Max streams allowed per connection.
-     * @param channelReleaser Method to release a channel and record on failure.
-     */
-    MultiplexedChannelRecord(Future<Channel> connectionFuture,
-                             long maxConcurrencyPerConnection,
-                             BiConsumer<Channel, MultiplexedChannelRecord> channelReleaser) {
-        this.connectionFuture = connectionFuture;
-        this.availableStreams = new AtomicLong(maxConcurrencyPerConnection);
-        this.childChannels = new ConcurrentHashMap<>(saturatedCast(maxConcurrencyPerConnection));
-        this.channelReleaser = channelReleaser;
-    }
+    // Only read or write in the connection.eventLoop()
+    private final Map<ChannelId, Http2StreamChannel> childChannels = new HashMap<>();
+
+    // Only write in the connection.eventLoop()
+    private volatile RecordState state = RecordState.OPEN;
 
-    @SdkTestInternalApi
-    MultiplexedChannelRecord(Future<Channel> connectionFuture,
-                             Channel connection,
-                             long maxConcurrencyPerConnection,
-                             BiConsumer<Channel, MultiplexedChannelRecord> channelReleaser) {
-        this.connectionFuture = connectionFuture;
-        this.childChannels = new ConcurrentHashMap<>(saturatedCast(maxConcurrencyPerConnection));
-        this.availableStreams = new AtomicLong(maxConcurrencyPerConnection);
-        this.channelReleaser = channelReleaser;
+    MultiplexedChannelRecord(Channel connection, long maxConcurrencyPerConnection) {
         this.connection = connection;
+        this.maxConcurrencyPerConnection = maxConcurrencyPerConnection;
+        this.availableChildChannels = new AtomicLong(maxConcurrencyPerConnection);
+    }
+
+    boolean acquireStream(Promise<Channel> promise) {
+        if (reserveStream()) {
+            releaseReservationOnFailure(promise);
+            acquireReservedStream(promise);
+            return true;
+        }
+        return false;
     }
 
-    MultiplexedChannelRecord acquire(Promise<Channel> channelPromise) {
-        availableStreams.decrementAndGet();
-        if (connection != null) {
-            createChildChannel(channelPromise);
-        } else {
-            connectionFuture.addListener((GenericFutureListener<Future<Channel>>) future -> {
-                if (future.isSuccess()) {
-                    connection = future.getNow();
-                    connection.attr(CHANNEL_POOL_RECORD).set(this);
-                    createChildChannel(channelPromise);
-                } else {
-                    channelPromise.setFailure(future.cause());
-                    channelReleaser.accept(connection, this);
+    private void acquireReservedStream(Promise<Channel> promise) {
+        doInEventLoop(connection.eventLoop(), () -> {
+            if (state != RecordState.OPEN) {
```

Comment:
It's not really synchronized, so it could happen at any time. I feel like we tried our best when we checked earlier. There's always going to be a race condition. I don't know if checking it again later helps very much.

Reply:
Since the connection updates the ping status and this runs in the same loop (I think), I don't think there's a race here

Reply:
I'm not sure I want to trust how the ping status is updated right now to guarantee correctness, since that could change in the future. If we want to guarantee correctness, I think we should add some kind of assurance that the ping status can only be updated in the event loop. I'd prefer just to have it be best-attempt. The time we got the read timeout is arbitrary. Whether or not we miss that it happened a few nanoseconds ago doesn't matter very much, because we already waited 30 seconds figure it out. What's another couple of nanoseconds?

Reply:
Okay that makes sense. I agree that this should be fine even if we miss the state change.